### PR TITLE
Resolves #62

### LINKED ISSUE #62: Unable to add additional parser arguments:
I have an entry-point that I want additional arguments, beyond those used by argschema:

```
> foo --version
```

```
> foo --input_json blah.json
```

Because the parse argument is instantiated in utils, and parsed immediately when ArgSchemaParse is constructed, I have no option to inject this arg.

### PR DESCRIPTION
**Pull request by nicain**
This allows me to inject a custom parser:

```
    parser = argparse.ArgumentParser()
    parser.add_argument('--version', action='version', version=__version__)

    """Main entry point for running pyproject_template."""
    mod = ArgSchemaParser(schema_type=InputParameters,
                          output_schema_type=OutputParameters,
                          parser=parser)
```

This results in:

```
➜ git:(74) ✗ [nicholasc@nicholasc-ubuntu]~/projects/aibs.pyproject_template/$aibs.pyproject_template --version
0.1.0
➜ git:(74) ✗ [nicholasc@nicholasc-ubuntu]~/projects/aibs.pyproject_template/$aibs.pyproject_template -h       
usage: aibs.pyproject_template [-h] [--version] [--output_json OUTPUT_JSON]
                               [--input_json INPUT_JSON]
                               [--log_level LOG_LEVEL]

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit

InputParameters:
  --output_json OUTPUT_JSON
                        file path to output json file
  --input_json INPUT_JSON
                        file path of input json file
  --log_level LOG_LEVEL
                        set the logging level of the module (default=ERROR)
```

All tox envs passing on my dev environment

### CODE REVIEW

**Comment by nicain** [comment]
@fcollman Let me know if this seems like something you want, and I can add a test to exercise and update the PR

**Comment by fcollman** [comment]
@nicain  just want to talk this through with you... why do this as opposed to extending the schema to include these additional arguments?

**Comment by nicain** [comment]
I want to injest --version, and use argparse action='version' to just print the version, and exit.  This way, my command-line script can use argschema to define its behavior, and I can also just ask "what version are you" and exit.

**Comment by fcollman** [comment]
So the only thing that gives me hesitation is that up to this point we haven’t explicitly advertised argparse as an interface for argschema. Meaning in the future we could move away from it as long as we maintain the same feature set without having the api change, and this would be extending the interface to explicitly allow Argparse classes in the api. There are alternative marshmallow hooks to allow post parse actions.  Seems that’s the feature you want to reproduce...

**Comment by nicain** [comment]
OK.

As long as I can digest schema-independent command-line args (--version, --dry-run are the two I want), I am happy  doing this with any tool.  

> There are alternative marshmallow hooks to allow post parse actions

I looked briefly for these, but couldn't find them.  Can you point me to some docs I can crib from?

**Comment by fcollman** [comment]
I’m traveling right now, but I’ll do that. Also I wonder if a refactor would make it easier to subclass ArgschemaParser with this functionality in a cleaner way.

**Comment by fcollman** [comment]
Also... check out #62. I’m not sure if we can merge these two ideas but you can see how this removes the input_json from the schema... is like to think of the right structure that lets loglevel also come out and maybe makes room for things like this.

I’m not sure whether you are married to the action=version functionality through argparse though.. 

A related functionality which is not provided now is to allow the command line flags like —version for booleans with no values. I think a PR that allows the schema to specify this behavior.

File: argschema/argschema_parser.py [hunk 1]
```diff
@@ -154,7 +155,7 @@ def __init__(self,
         self.logger.debug('input_data is {}'.format(input_data))
 
         # convert schema to argparse object
-        p = utils.schema_argparser(self.schema)
+        p = utils.schema_argparser(self.schema, parser=parser)
```
Comment (fcollman) [inline argschema/argschema_parser.py]: what if we made a method on ArgSchemaParser of 

```
def get_parser(self):
   return utils.schema_argparser(self.schema)
```

and that way extending ArgSchemaParser would be cleaner.

**Comment by fcollman** [comment]
what do you think about this?

```
def ArgSchemaParserPlus(ArgSchemaParser):
    
    def __init__(self,parser=None,*args,**kwargs):
        [argsA, extrasA] = parser.parse_known_args()
        super(ArgSchemaParserPlus,self).__init__(args=extrasA,*args,**kwargs)
```

**Comment by fcollman** [comment]
actually.. i think this is better

```
def ArgSchemaParserPlus(ArgSchemaParser):
    
    def __init__(self,*args,**kwargs):
        args = kwargs.pop('args')
        [argsA, extrasA] = parser.parse_known_args(args=args)
        super(ArgSchemaParserPlus,self).__init__(args=extrasA,*args,**kwargs)
```

**Comment by nicain** [comment]
It took some iteration, but this works:

```
class ArgSchemaParserPlus(ArgSchemaParser):
    
    def __init__(self,*args,**kwargs):

        parser = argparse.ArgumentParser() 
        parser.add_argument('--version', action='version', version=__version__)
        [known_args, extra_args] = parser.parse_known_args()
        self.args = known_args 

        super(ArgSchemaParserPlus,self).__init__(args=extra_args, **kwargs)
```

**Pull request closed without merging**